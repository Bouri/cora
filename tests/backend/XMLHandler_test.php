<?php 
/*
 * Copyright (C) 2015-2017 Marcel Bollmann <bollmann@linguistics.rub.de>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */ ?>
<?php
require_once"data/test_data.php";
require_once"mocks/LocaleHandler_mock.php";
require_once"{$GLOBALS['CORA_WEB_DIR']}/lib/xmlHandler.php";

/** A mock DBInterface to trick the XMLHandler with
 */
class Cora_Tests_DBInterface_for_XMLHandler_Mock {
    /// the imported document and options, we'll run asserts on these
    public $document;
    public $options;

    // expected and initial data
    private $expected;
    private $initial;

    function __construct() {
        $this->expected = get_XMLHandler_expected();
        $this->initial = get_XMLHandler_initial();
    }

    public function insertNewDocument($options, $data) {
        $this->document = $data;
        $this->options = $options;
        return array('success' => true, 'warnings' => array());
    }
    public function getAllLines($fileid) {
        return $this->initial["lines"];
    }
    public function getAllSuggestions($fileid, $lineid) {
        // XXX needed for export
        return array(
        );
    }
    public function openFile($fileid) {
        if ($fileid == "512") {
            return array("success" => false);
        }

        return array(
            "lastEditedRow" => "",
            "data" => array(
                "id" => "",
                "sigle" => "",
                "fullname" => "",
                "project_id" => "",
                "created" => "",
                "creator_id" => "",
                "changed" => "",
                "changer_id" => "",
                "currentmod_id" => "",
                "ext_id" => "",
                "file_name" => "bla",
                "tagset" => "",
                "header" => $this->expected["header"]
            ),
            "success" => true
        );
    }
}

class Cora_Tests_XMLHandler_test extends PHPUnit_Framework_TestCase {
    protected $dbi;
    protected $xh;
    protected $test_data;

    protected function setUp() {
        $this->test_data = get_XMLHandler_expected();
        $this->dbi = new Cora_Tests_DBInterface_for_XMLHandler_Mock();
        $this->xh = new XMLHandler($this->dbi, new MockLocaleHandler());

        $options = array('sigle' => 't1', 'name' => 'testdocument');
        $test_doc_data = get_CoraDocument_data();
        $this->cd = new CoraDocument($options, new MockLocaleHandler());
        $this->cd->setLayoutInfo($test_doc_data["pages"],
                                 $test_doc_data["columns"],
                                 $test_doc_data["lines"]);
        $this->cd->setTokens($test_doc_data["tokens"],
                             $test_doc_data["dipls"],
                             $test_doc_data["mods"]);
        $this->cd->setShiftTags($test_doc_data["shifttags"]);
        $this->cd->setComments($test_doc_data["comments"]);
    }

    public function testImport() {
        $options = array('ext_id' => '');
        $filename = array(
            "tmp_name" => __DIR__ . "/data/cora-importtest.xml",
            "name" => __DIR__ . "/cora-importtest.xml"
        );
        $this->xh->import($filename, $options, 1);

        $this->assertEquals($this->test_data["options"]["ext_id"],
                            $this->dbi->options["ext_id"]);
        $this->assertStringEndsWith($this->test_data["options"]["name"],
                                    $this->dbi->options["name"]);

        $this->assertEquals($this->test_data["tokens"],
                            $this->dbi->document->getTokens());

        $this->assertEquals($this->test_data["moderns"],
                            $this->dbi->document->getModerns());

        $this->assertEquals($this->test_data["dipls"],
            $this->dbi->document->getDipls());

        $this->assertEquals($this->test_data["lines"],
            $this->dbi->document->getLines());

        $this->assertEquals($this->test_data["columns"],
            $this->dbi->document->getColumns());

        $this->assertEquals($this->test_data["pages"],
            $this->dbi->document->getPages());
        $this->assertEquals($this->test_data["shifttags"],
            $this->dbi->document->getShifttags());
        $this->assertEquals($this->test_data["comments"],
            $this->dbi->document->getComments());

        // currently, the following assert fails since it imports the padding
        // whitespace
        $this->assertEquals($this->test_data["header"],
            $this->dbi->document->getHeader());
    }

    public function testExport() {
        $dom = $this->xh->serializeDocument($this->cd);
        $this->assertNotEmpty($dom);
        $xml_string = $dom->saveXML();
        $this->assertEquals("<", substr($xml_string, 0, 1));
        libxml_use_internal_errors(true);
        $xml_dom = new DOMDocument();
        $xml_dom->loadXML($xml_string);
        $this->assertNotFalse($xml_dom);
        $this->assertTrue($xml_dom->relaxNGValidate($GLOBALS["CORA_RELAXNG"]));
        $this->assertEmpty(libxml_get_errors());
        $this->assertXmlStringEqualsXmlFile(__DIR__ . '/data/XMLHandler_test_expected.xml',
                                            $xml_string);
    }
}
?>
