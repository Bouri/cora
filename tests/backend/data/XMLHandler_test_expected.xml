<text id="t1">
  <cora-header name="testdocument" sigle="t1"/>
  <header/>
  <layoutinfo>
    <page id="p1" no="42" range="c1" side="v"/>
    <column id="c1" name="a" range="l1..l2"/>
    <line id="l1" name="01" range="t1_d1..t2_d1"/>
    <line id="l2" name="02" range="t2_d2..t3_d1"/>
  </layoutinfo>
  <shifttags>
    <title range="t3..t3"/>
  </shifttags>
  <token id="t1" trans="$ol|tu">
    <dipl id="t1_d1" trans="$ol|tu" utf=""/>
    <mod ascii="" checked="y" id="t1_m1" trans="$ol|" utf="">
      <norm tag="sollst"/>
      <suggestions>
        <pos score="0.91" source="user" tag="VVFIN"/>
        <lemma source="user" tag="sollen"/>
        <morph score="0.91" source="user" tag="2.Sg.Pres.Ind"/>
      </suggestions>
    </mod>
    <mod ascii="" checked="n" id="t1_m2" trans="tu" utf="">
      <norm tag="du"/>
      <comment tag="Interessantes PhÃ¤nomen hier"/>
      <suggestions>
        <pos score="0.91" source="user" tag="PPER"/>
        <lemma source="user" tag="du"/>
        <morph score="0.91" source="user" tag="2.Sg.*.Nom"/>
      </suggestions>
      <cora-flag name="boundary"/>
    </mod>
  </token>
  <token id="t2" trans="ge#e$$en">
    <dipl id="t2_d1" trans="ge#" utf=""/>
    <dipl id="t2_d2" trans="e$$en" utf=""/>
    <mod ascii="" checked="n" id="t2_m1" trans="ge#e$$en" utf="">
      <norm tag="gegessen"/>
      <suggestions>
        <pos score="0.91" source="user" tag="VVPP"/>
        <lemma source="user" tag="essen"/>
      </suggestions>
    </mod>
  </token>
  <comment type="K">yay foo bar whoo</comment>
  <token id="t3" trans="Anshelm/(.)">
    <dipl id="t3_d1" trans="Anshelm/(.)" utf=""/>
    <mod ascii="" checked="n" id="t3_m1" trans="Anshelm" utf="">
      <norm tag="Anselm"/>
      <suggestions>
        <pos score="0.91" source="user" tag="NE"/>
        <lemma source="user" tag="Anselm"/>
        <morph score="0.91" source="user" tag="Masc.Nom.Sg"/>
      </suggestions>
    </mod>
    <mod ascii="" checked="n" id="t3_m2" trans="/" utf="">
      <norm tag="/"/>
      <suggestions>
        <pos score="0.91" source="user" tag="$_"/>
        <lemma source="user" tag="/"/>
      </suggestions>
      <cora-flag name="general error"/>
    </mod>
    <mod ascii="" checked="n" id="t3_m3" trans="(.)" utf="">
      <norm tag="."/>
      <suggestions>
        <pos score="0.91" source="user" tag="$."/>
        <lemma source="user" tag="."/>
      </suggestions>
      <cora-flag name="general error"/>
      <cora-flag name="boundary"/>
    </mod>
  </token>
</text>
